lib/Sema/NameBinding.cpp

```diff
@@ -297,6 +298,551 @@ static void insertPrecedenceGroupDecl(NameBinder &binder, SourceFile &SF,
+        BraceStmt *NB = transformBraceStmt(B);
+        if (NB != B) {
+          DCS->setBody(NB);
+        }
+      }
+      for (CatchStmt *C : DCS->getCatches()) {
+        if (BraceStmt *CB = dyn_cast_or_null<BraceStmt>(C->getBody())) {
+          BraceStmt *NCB = transformBraceStmt(CB);
+          if (NCB != CB) {
+            C->setBody(NCB);
+          }
+        }
+      }
+      return DCS;
+    }
+
+    BraceStmt *transformBraceStmt(BraceStmt *BS) {
+      ASTContext &Context = Binder.Context;
+      SmallVector<ASTNode, 3> Elements;
+      for (auto &elt : BS->getElements()) {
+        // If this is an expression we have to walk it to transform any interior
+        // closure expression bodies.
+        if (Expr *E = elt.dyn_cast<Expr *>()) {
+          Elements.push_back(E->walk(CF));
+          continue;
+        }
+
+        // We only transform statements.
+        Stmt *stmt = elt.dyn_cast<Stmt *>();
+        if (!stmt) {
+          Elements.push_back(elt);
+          continue;
+        }
+
+        // Build configurations are handled later.  If this isn't a build
+        // configuration transform the statement only.
+        IfConfigStmt *config = dyn_cast<IfConfigStmt>(stmt);
+        if (!config) {
+          Elements.push_back(transformStmt(stmt));
+          continue;
+        }
+
+        for (auto &clause : config->getClauses()) {
+          // Evaluate conditions until we find the active clause.
+          if (clause.Cond && !evaluateCondition(clause.Cond)) {
+            continue;
+          }
+
+          // Now that we have the active clause, look into its elements for
+          // further statements to transform.
+          for (auto &clauseElt : clause.Elements) {
+            auto innerStmt = clauseElt.dyn_cast<Stmt *>();
+            if (!innerStmt) {
+              Elements.push_back(clauseElt);
+              continue;
+            }
+
+            auto innerConfig = dyn_cast<IfConfigStmt>(innerStmt);
+            if (!innerConfig) {
+              Elements.push_back(transformStmt(innerStmt));
+              continue;
+            }
+
+            // Nested build configurations require an additional transform to
+            // evaluate completely.
+            auto BS = BraceStmt::create(Context, SourceLoc(),
+                                        {innerConfig}, SourceLoc());
+            for (auto &es : transformBraceStmt(BS)->getElements()) {
+              if (auto innerInnerStmt = es.dyn_cast<Stmt *>()) {
+                Elements.push_back(transformStmt(innerInnerStmt));
+              } else {
+                Elements.push_back(es);
+              }
+            }
+          }
+          break;
+        }
+      }
+
+      return swift::BraceStmt::create(Context, BS->getLBraceLoc(),
+                                      Context.AllocateCopy(Elements),
+                                      BS->getRBraceLoc());
+    }
+
+    BraceStmt *resolveClausesAndExtractTLCDs(BraceStmt *BS) {
```

Comment:
A *lot* of this is the same as the normal `transformBraceStmt`. Think there's a way to factor out the common parts? Maybe with `llvm::function_ref` for callbacks?